beacon_node/execution_layer/src/lib.rs

```diff
@@ -1404,18 +1451,183 @@ impl<T: EthSpec> ExecutionLayer<T> {
+            InvalidBuilderPayload::Fork { payload, expected } => {
+                write!(f, "payload fork was {:?} not {}", payload, expected)
+            }
+            InvalidBuilderPayload::Signature { signature, pubkey } => write!(
+                f,
+                "invalid payload signature {} for pubkey {}",
+                signature, pubkey
+            ),
+        }
+    }
+}
+
+/// Perform some cursory, non-exhaustive validation of the bid returned from the builder.
+fn verify_builder_bid<T: EthSpec, Payload: ExecPayload<T>>(
+    bid: &ForkVersionedResponse<SignedBuilderBid<T, Payload>>,
+    parent_hash: ExecutionBlockHash,
+    prev_randao: Hash256,
+    timestamp: u64,
+    block_number: Option<u64>,
+    profit_threshold: Uint256,
+    spec: &ChainSpec,
+) -> Result<(), Box<InvalidBuilderPayload>> {
+    let is_signature_valid = bid.data.verify_signature(spec);
+    let header = &bid.data.message.header;
+
+    let payload_value = bid.data.message.value;
+    if payload_value < profit_threshold {
+        Err(Box::new(InvalidBuilderPayload::LowValue {
+            profit_threshold,
+            payload_value,
+        }))
+    } else if header.parent_hash() != parent_hash {
+        Err(Box::new(InvalidBuilderPayload::ParentHash {
+            payload: header.parent_hash(),
+            expected: parent_hash,
+        }))
+    } else if header.prev_randao() != prev_randao {
+        Err(Box::new(InvalidBuilderPayload::PrevRandao {
+            payload: header.prev_randao(),
+            expected: prev_randao,
+        }))
+    } else if header.timestamp() != timestamp {
+        Err(Box::new(InvalidBuilderPayload::Timestamp {
+            payload: header.timestamp(),
+            expected: timestamp,
+        }))
+    } else if block_number.map_or(false, |n| n != header.block_number()) {
+        Err(Box::new(InvalidBuilderPayload::BlockNumber {
+            payload: header.block_number(),
+            expected: block_number,
+        }))
+    } else if !matches!(bid.version, Some(ForkName::Merge)) {
+        // Once fork information is added to the payload, we will need to
+        // check that the local and relay payloads match. At this point, if
+        // we are requesting a payload at all, we have to assume this is
+        // the Bellatrix fork.
+        Err(Box::new(InvalidBuilderPayload::Fork {
+            payload: bid.version,
+            expected: ForkName::Merge,
+        }))
+    } else if !is_signature_valid {
+        Err(Box::new(InvalidBuilderPayload::Signature {
+            signature: bid.data.signature.clone(),
+            pubkey: bid.data.message.pubkey,
+        }))
+    } else {
+        Ok(())
+    }
+}
+
+/// A helper function to record the time it takes to execute a future.
+async fn timed_future<F: Future<Output = T>, T>(metric: &str, future: F) -> (T, Duration) {
+    let start = Instant::now();
+    let result = future.await;
+    let duration = Instant::now().duration_since(start);
```

Comment:
Equivalently:

```suggestion
    let duration = start.elapsed();
```

(https://doc.rust-lang.org/std/time/struct.Instant.html#method.elapsed)